SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityUsagePoller.java

```diff
@@ -190,4 +206,88 @@ private void updateLongRunningTasksUsage(Map<ResourceUsageType, Number> longRunn
     longRunningTasksUsage.compute(ResourceUsageType.CPU_USED, (k, v) -> (v == null) ? cpuUsed : v.doubleValue() + cpuUsed);
   }
+
+  private SingularityClusterUtilization getClusterUtilization(Map<String, RequestUtilization> utilizationPerRequestId, double minUnderUtilizedPct, long totalMemBytesUsed, long totalMemBytesAvailable, double totalCpuUsed, double totalCpuAvailable, long now) {
+    int numRequestsWithUnderUtilizedCpu = 0;
+    int numRequestsWithOverUtilizedCpu = 0;
+    int numRequestsWithUnderUtilizedMemBytes = 0;
+
+    double totalUnderUtilizedCpu = 0;
+    double totalOverUtilizedCpu = 0;
+    long totalUnderUtilizedMemBytes = 0;
+
+    double maxUnderUtilizedCpu = Double.MIN_VALUE;
+    double maxOverUtilizedCpu = Double.MIN_VALUE;
+    long maxUnderUtilizedMemBytes = Long.MIN_VALUE;
+
+    double minUnderUtilizedCpu = Double.MAX_VALUE;
+    double minOverUtilizedCpu = Double.MAX_VALUE;
+    long minUnderUtilizedMemBytes = Long.MAX_VALUE;
+
+
+    for (Iterator<Entry<String, RequestUtilization>> it = utilizationPerRequestId.entrySet().iterator(); it.hasNext(); ) {
+      RequestUtilization utilization = it.next().getValue();
+      Optional<SingularityDeploy> maybeDeploy = deployManager.getDeploy(utilization.getRequestId(), utilization.getDeployId());
+
+      if (maybeDeploy.isPresent() && maybeDeploy.get().getResources().isPresent()) {
+        boolean includeUtilization = true;
+        long memoryBytesReserved = (long) (maybeDeploy.get().getResources().get().getMemoryMb() * SingularitySlaveUsage.BYTES_PER_MEGABYTE);
+        double cpuReserved = maybeDeploy.get().getResources().get().getCpus();
+
+        double unusedCpu = cpuReserved - utilization.getAvgCpuUsed();
+        long unusedMemBytes = memoryBytesReserved - utilization.getMemBytesTotal();
+
+        if (unusedCpu / cpuReserved >= minUnderUtilizedPct) {
+          numRequestsWithUnderUtilizedCpu++;
+          totalUnderUtilizedCpu += unusedCpu;
+          maxUnderUtilizedCpu = Math.max(unusedCpu, maxUnderUtilizedCpu);
+          minUnderUtilizedCpu = Math.min(unusedCpu, minUnderUtilizedCpu);
+        } else if (unusedCpu < 0) {
+          double overusedCpu = Math.abs(unusedCpu);
+
+          numRequestsWithOverUtilizedCpu++;
+          totalOverUtilizedCpu += overusedCpu;
+          maxOverUtilizedCpu = Math.max(overusedCpu, maxOverUtilizedCpu);
+          minOverUtilizedCpu = Math.min(overusedCpu, minOverUtilizedCpu);
+        } else {
+          includeUtilization = false;
+        }
+
+        if (unusedMemBytes / memoryBytesReserved >= minUnderUtilizedPct) {
+          numRequestsWithUnderUtilizedMemBytes++;
+          totalUnderUtilizedMemBytes += unusedMemBytes;
+          maxUnderUtilizedMemBytes = Math.max(unusedMemBytes, maxUnderUtilizedMemBytes);
+          minUnderUtilizedMemBytes = Math.min(unusedMemBytes, minUnderUtilizedMemBytes);
+        } else if (!includeUtilization) {
+          it.remove();
```

Comment:
yeah, i wasn't a fan of the iterator here either, but that was the only way to safely remove the item from the list in the loop. If we decide to keep the min percentage, I could do a second cleanup loop rather than do it within the loop